allennlp/semparse/contexts/atis_tables.py

```diff
@@ -65,31 +120,49 @@ def get_date_from_utterance(tokenized_utterance: List[Token],
         return datetime(year, month, day)
     return None
 
-def get_numbers_from_utterance(utterance: str) -> List[str]:
+def get_numbers_from_utterance(utterance: str, tokenized_utterance: List[Token]) -> Dict[str, List[int]]:
     """
-    Given an utterance, find all the numbers that are in the action space.
+    Given an utterance, this function finds all the numbers that are in the action space. Since we need to
+    keep track of linking scores, we represent the numbers as a dictionary, where the keys are the string
+    representation of the number and the values are lists of the token indices that triggers that number.
     """
-    numbers = []
-    numbers.extend(re.findall(r'\d+', utterance))
-    numbers.extend([str(int(num_str.rstrip('pm')) * HOUR_TO_TWENTY_FOUR + TWELVE_TO_TWENTY_FOUR)
-                    for num_str in re.findall(r'\d+', utterance)])
-
-    numbers.extend(get_times_from_utterance(utterance))
-
-    words = utterance.split(' ')
-    for word in words:
-        if word in MONTH_NUMBERS:
-            numbers.append(str(MONTH_NUMBERS[word]))
-        if word in DAY_NUMBERS:
-            numbers.append(str(DAY_NUMBERS[word]))
-        if word in MISC_TIME_TRIGGERS:
-            numbers.extend(MISC_TIME_TRIGGERS[word])
-
-    for tens, digits in zip(words, words[1:]):
-        day = ' '.join([tens, digits])
-        if day in DAY_NUMBERS:
-            numbers.append(str(DAY_NUMBERS[day]))
-    return sorted(numbers, reverse=True)
+
+    # When we use a regex to find numbers or strings, we need a mapping from
+    # the character to which token triggered it.
+    char_offset_to_token_index = {token.idx : token_idx for token_idx, token
+                                  in enumerate(tokenized_utterance)}
+
+    # We want to look up later for each time whether it appears after a word
+    # such as "about" or "approximately".
+    indices_approximate_words = [idx for idx, token in enumerate(tokenized_utterance) \
+            if token.text in APPROX_WORDS]
+    number_linking_dict = get_regex_match(r'\d+',
+                                          utterance,
+                                          char_offset_to_token_index,
+                                          lambda match: [int(match)],
+                                          indices_approximate_words)
+
+    times_linking_dict = get_times_from_utterance(utterance,
+                                                  char_offset_to_token_index,
+                                                  indices_approximate_words)
+
+    for key, value in times_linking_dict.items():
+        number_linking_dict[key].extend(value)
+
+    for idx, token in enumerate(tokenized_utterance):
+        if token.text in MONTH_NUMBERS:
```

Comment:
It probably makes sense to make a `NUMBER_TRIGGER_DICT` similar to your `ATIS_TRIGGER_DICT`, so you don't have to have so many independent-but-similar checks here.

Reply:
This reason I did not combine these here is because we want to actually construct the ``datetime`` object and use that to look up what day of the week later. To do that we need to know what value `month` and `day` are. If we merge these together, we can't do ``month = ...`` since it's a generic numbers dictionary.

Reply:
I'm confused; can you point me to where you're constructing this `datetime` object?  That's not happening here, right?  You're just returning a mapping from `str` to `List[int]`.

Reply:
Oh sorry, you're right! I thought this was a comment on ``get_date_from_utterance``.